planner/cascades/transformation_rules.go

```diff
@@ -589,3 +592,43 @@ func (r *PushSelDownJoin) OnTransform(old *memo.ExprIter) (newExprs []*memo.Grou
 	newJoinExpr.SetChildren(leftGroup, rightGroup)
 	return []*memo.GroupExpr{newJoinExpr}, true, false, nil
 }
+
+// PushTopNDownProjection pushes TopN to Projection.
+type PushTopNDownProjection struct {
+	baseRule
+}
+
+// NewRulePushTopNDownProjection creates a new Transformation PushTopNDownProjection.
+func NewRulePushTopNDownProjection() Transformation {
+	rule := &PushTopNDownProjection{}
+	rule.pattern = memo.BuildPattern(
+		memo.OperandTopN,
+		memo.EngineTiDBOnly,
+		memo.NewPattern(memo.OperandProjection, memo.EngineTiDBOnly),
+	)
+	return rule
+}
+
+// OnTransform implements Transformation interface.
+// This rule tries to pushes the TopN through Projection.
+func (r *PushTopNDownProjection) OnTransform(old *memo.ExprIter) (newExprs []*memo.GroupExpr, eraseOld bool, eraseAll bool, err error) {
+	topN := old.GetExpr().ExprNode.(*plannercore.LogicalTopN)
+	proj := old.Children[0].GetExpr().ExprNode.(*plannercore.LogicalProjection)
+	childGroup := old.Children[0].GetExpr().Children[0]
+	for _, by := range topN.ByItems {
+		by.Expr = expression.ColumnSubstitute(by.Expr, old.Children[0].Group.Prop.Schema, proj.Exprs)
+	}
+	// remove meaningless constant sort items.
+	for i := len(topN.ByItems) - 1; i >= 0; i-- {
+		switch topN.ByItems[i].Expr.(type) {
+		case *expression.Constant, *expression.CorrelatedColumn:
+			topN.ByItems = append(topN.ByItems[:i], topN.ByItems[i+1:]...)
+		}
+	}
```

Comment:
We'd better not modify the old TopN.
Create a new TopN here, clone the old one's `ByItems` and then use `ColumnSubstitue` on the new TopN.

Reply:
`ColumnSubsitute` is not in-place change. Just create a new TopN is enough, no need to clone it.

Reply:
Do you mean we don't have to clone ByItems? Using the original topN's ByItems? Actually, I've already updated. Could you please have a look at the latest change?